clowder-framework/clowder
# Extractor Catalog: search filters + matching ANY/ALL

## Description

### Problem
Design includes more ways to filter catalog results than we currently offer.

### Approach
* Add generic search filter to the Extractor Catalog
* Add space search filter to the Extractor Catalog
* Add metadata search filter to the Extractor Catalog
    * NOTE: this assumes `ExtractorInfo.contexts` has been populated properly.. for example, the wordcount extractor contains the proper entries
* Option to match either ALL filters or ANY filter
* Make sure these new filters play nicely with the label filter

### How to Test
Prerequisites: RabbitMQ plugin enabled, at least two different types of extractors registered, wordcount extractor registered

1. Checkout and run this branch locally
2. Login to Clowder
3. Navigate to the Extractor Catalog
    * You should see a text input allowing for a general text search (e.g. name, authors, description, etc)
    * You should see radio buttons offering to "match ANY" or "match ALL"
4. Enter a nonsensical combination, for example: `Match ALL` and a general text search of `asdf`
    * You should see no results
5. Choose a label (e.g. Image), `Match ALL`, and general search of "Rob Kooper"
    * You should see only image-based extractors that include "Rob Kooper" as a contributor or author
6. Choose no label, enter "Image" as generic search with `Match ANY` and choose "words" from the metadata dropdown
    * You should be shown any extractors mentioning "Image", as well as the wordcount extractor
7. Same as 6, except choose `Match ALL` this time
    * You should see no results

NOTE: If a Label is selected, then we calculate its intersect with the results of the other filters regardless of whether "match ANY" / "match ALL" is selected. This is to avoid the case where the user chooses "Image" explicitly, where "match ANY" will can non-image results to that list.

## Review Time Estimate


## Types of changes


## Checklist:
